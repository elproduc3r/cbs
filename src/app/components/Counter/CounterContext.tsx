import { createContext } from "react";

interface CounterContextType {
  pageVisits: number;
}

export const CounterContext = createContext<CounterContextType>({
  pageVisits: 1
});