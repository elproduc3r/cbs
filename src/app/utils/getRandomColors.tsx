/**
 * 
 * @param num  number of colors (size of array) to return
 * @returns array of colors that has length of num
 */

const getRandomColors = (num: number): string[] => {
  const COLORS = ["aliceblue","antiquewhite","aqua","aquamarine","azure","beige","bisque",
    "black","blanchedalmond","blue","blueviolet","brown","burlywood","cadetblue","chartreuse",
    "chocolate","coral","cornflowerblue","cornsilk","crimson","cyan","darkblue","darkcyan",
    "darkgoldenrod","darkgray","darkgreen","darkgrey","darkkhaki","darkmagenta",
    "darkolivegreen","darkorange","darkorchid","darkred","darksalmon","darkseagreen",
    "darkslateblue","darkslategray","darkslategrey","darkturquoise","darkviolet","deeppink",
    "deepskyblue","dimgray","dimgrey","dodgerblue","firebrick","floralwhite","forestgreen",
    "fuchsia","gainsboro","ghostwhite","gold","goldenrod","gray","green","greenyellow","grey",
    "honeydew","hotpink","indianred","indigo","ivory","khaki","lavender","lavenderblush"
  ];
  const getRandomNum = ():number => (Math.floor(Math.random() * (COLORS.length - 1)) + 0);
  const result = [COLORS[getRandomNum()]];
  while(result.length < num) {
    const nextRandomNum = getRandomNum();
    const nextColor = COLORS[nextRandomNum];
    if(result.indexOf(nextColor) < 0) {
      result.push(nextColor)
    }
  }
  return result;
};

export default getRandomColors;