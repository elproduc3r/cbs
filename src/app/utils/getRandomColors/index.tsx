/**
 * 
 * @param num  number of colors (size of array) to return
 * @returns array of colors that has length of num
 */

const getRandomColors = (num: number): string[] => {
  const COLORS = [
    'black','blue','brown','chocolate','coral','crimson','darkblue','darkcyan',
    'darkgoldenrod','darkgray','darkgreen','darkkhaki','darkmagenta','darkolivegreen',
    'darkorange','darkorchid','darkred','darksalmon','darkslateblue','darkslategray',
    'darkturquoise','darkviolet','deeppink','firebrick','forestgreen','fuchsia',
    'green','indigo','maroon','mediumblue','mediumorchid','mediumseagreen',
    'mediumslateblue','mediumvioletred', 'navy','olive','olivedrab', 'orangered',
    'purple','red','saddlebrown','salmon','sienna','slateblue','slategray', 'steelblue',
    'teal','tomato','violet'
  ];
  const getRandomNum = ():number => (Math.floor(Math.random() * (COLORS.length - 1)) + 0);
  const result = [COLORS[getRandomNum()]];
  while(result.length < num) {
    const nextColor = COLORS[getRandomNum()];
    if(result.indexOf(nextColor) < 0) {
      result.push(nextColor);
    }
  }
  return result;
};

export default getRandomColors;
