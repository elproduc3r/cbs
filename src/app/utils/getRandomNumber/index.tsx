export const getRandomNum = (min: number, max: number) => Math.floor(Math.random() * max) + min;