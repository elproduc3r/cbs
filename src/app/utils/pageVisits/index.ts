export interface CBSLSDataType {
  pageVisits: number;
};

export const CBS_KEY = 'cbs';

/**
 * On initial page load, this should increment the pageVisits.
 * 
 * If there is no data, it will default to zero. Then pageVisits
 * is incremented by 1 and updated in localStorage.
 * 
 * If the optional reset argument is true, the new pageVisit is overriden
 * and set to 1.
 * 
 * @returns the new pageVisit value which should be set in Context Provider
 */
export const initializePageVisits = (reset?: boolean): number => {
  const lsData = localStorage.getItem(CBS_KEY);
  let cbsData: CBSLSDataType = lsData ? JSON.parse(lsData) : {pageVisits: 0};
  cbsData.pageVisits = reset ? 1 : cbsData.pageVisits + 1;
  localStorage.setItem(CBS_KEY, JSON.stringify(cbsData));
  return cbsData.pageVisits;
};

/**
 * 
 * @returns the value for pageVisits that is stored in localStorage
 */
export const getPageVisits = (): number => {
  const lsData = localStorage.getItem(CBS_KEY);
  let cbsData: CBSLSDataType = lsData ? JSON.parse(lsData) : {pageVisits: 1};
  return cbsData.pageVisits;
}
