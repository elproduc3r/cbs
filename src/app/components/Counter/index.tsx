'use client'

import React, {useEffect, useState} from 'react';
import { CounterContext } from './CounterContext';
import { initializePageVisits } from '../../utils/pageVisits';

export interface CounterProps {
  children: React.ReactNode;
};

const Counter = ({
  children
}: CounterProps): JSX.Element => {
  const [pageVisits , setPageVisits] = useState(1);
  useEffect(() => {
    const pageVisits = initializePageVisits();
    setPageVisits(pageVisits);
  }, []);

  return (
    <CounterContext.Provider value={{pageVisits}}>
      {children}
    </CounterContext.Provider>
  )
};

export default Counter;
export {CounterContext};