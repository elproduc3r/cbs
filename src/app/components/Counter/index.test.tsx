import React from 'react';
import { render, screen } from '@testing-library/react';
import { CBS_KEY } from '../../utils/pageVisits';
import Counter from '.';


interface StoreType {
  [index: string]: string;
};

var localStorageMock = (function() {
  var store: StoreType = {};
  return {
    getItem: function(key: string) {
      return store[key];
    },
    setItem: function(key: string, value: string) {
      store[key] = value.toString();
    },
    clear: function() {
      store = {};
    },
    removeItem: function(key: string) {
      delete store[key];
    }
  };
})();
Object.defineProperty(window, 'localStorage', { value: localStorageMock });

describe("<Counter>", () => {
  test('snapshot', () => {
    const {container} = render(
      <Counter>
        <div>child text</div>
      </Counter>
    );
    expect(container).toMatchSnapshot();
  });
  test('renders children', () => {
    const childText = "child component";
    render(
      <Counter>
        <div>{childText}</div>
      </Counter>
    );
    const textElement = screen.getByText(childText);
    expect(textElement).toBeInTheDocument();
  });
  describe("localStorage", () => {
    test('updates correct value in localStorage when NO data exists', () => {
      localStorage.clear();
      render(
        <Counter>
          <div>child</div>
        </Counter>
      );
      let lsData = localStorage.getItem(CBS_KEY);
      let jsonData = lsData ? JSON.parse(lsData) : {pageVisits: 1};
      expect(jsonData.pageVisits).toEqual(1);
    });
    test('updates correct value in localStorage when data exists', () => {
      const mockData = {
        pageVisits: 3
      };
      localStorage.setItem(CBS_KEY, JSON.stringify(mockData));
      render(
        <Counter>
          <div>child</div>
        </Counter>
      );
      let lsData = localStorage.getItem(CBS_KEY);
      let jsonData = lsData ? JSON.parse(lsData) : {pageVisits: 1};
      expect(jsonData.pageVisits).toEqual(4);
    });
  });
});
