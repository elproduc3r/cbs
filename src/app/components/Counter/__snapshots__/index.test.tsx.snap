// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter> snapshot 1`] = `
<div>
  <div>
    child text
  </div>
</div>
`;
